File: content/snippets/python/interview_lcci/02/01_remove_duplicate_node.py
class ListNode:
    def __init__(self, val: int, next: "ListNode | None" = None) -> None:
        self.val = val
        self.next = next


def build_list(nums: list[int]) -> ListNode | None:
    dummy = ListNode(0)
    tail = dummy
    for n in nums:
        tail.next = ListNode(n)
        tail = tail.next
    return dummy.next


def remove_duplicate_nodes(head: ListNode | None) -> None:
    seen: set[int] = set()
    prev: ListNode | None = None
    cur = head

    while cur is not None:
        if cur.val in seen:
            assert prev is not None
            prev.next = cur.next
            cur = cur.next
            continue
        seen.add(cur.val)
        prev = cur
        cur = cur.next


def print_list(head: ListNode | None) -> None:
    out: list[str] = []
    while head is not None:
        out.append(str(head.val))
        head = head.next
    print(" ".join(out))


def main() -> None:
    head = build_list([1, 2, 3, 3, 2, 1, 4])
    remove_duplicate_nodes(head)
    print_list(head)


if __name__ == "__main__":
    main()

File: content/snippets/typescript/interview_lcci/02/01_remove_duplicate_node.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function buildList(nums: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let tail = dummy;
  for (const n of nums) {
    tail.next = new ListNode(n);
    tail = tail.next;
  }
  return dummy.next;
}

function removeDuplicateNodes(head: ListNode | null): void {
  const seen = new Set<number>();
  let prev: ListNode | null = null;
  let cur = head;

  while (cur !== null) {
    if (seen.has(cur.val)) {
      if (prev !== null) {
        prev.next = cur.next;
      }
      cur = cur.next;
      continue;
    }

    seen.add(cur.val);
    prev = cur;
    cur = cur.next;
  }
}

function printList(head: ListNode | null): void {
  const out: number[] = [];
  while (head !== null) {
    out.push(head.val);
    head = head.next;
  }
  console.log(out.join(" "));
}

function main(): void {
  const head = buildList([1, 2, 3, 3, 2, 1, 4]);
  removeDuplicateNodes(head);
  printList(head);
}

main();
